hawkular-alerts-engine-extensions/hawkular-alerts-events-aggregation/src/main/java/org/hawkular/alerts/extensions/EventsAggregationExtension.java

```diff
@@ -106,68 +109,125 @@ public void init() {
                 ENGINE_EXTENSIONS_DEFAULT));
         defaultExpiration = properties.getProperty(EVENTS_EXPIRATION, EVENTS_EXPIRATION_ENV, EVENTS_EXTENSIONS_DEFAULT);
         if (engineExtensions) {
-            log.info("Registering Trigger CREATE/UPDATE/REMOVE listener");
-            definitions.registerListener(events -> refresh(events), TRIGGER_CREATE, TRIGGER_UPDATE, TRIGGER_REMOVE);
+            log.info("Registering Distributed Trigger listener");
+            definitions.registerDistributedListener(events -> refresh(events));
             extensions.addExtension(this);
         }
     }
 
-    private void refresh(Set<DefinitionsEvent> defEvents) {
-        boolean refresh = false;
-        for (DefinitionsEvent defEvent : defEvents) {
-            if (defEvent.getTags() != null && TAG_VALUE.equals(defEvent.getTags().get(TAG_NAME))) {
-                refresh = true;
-                break;
-            }
+    private class TriggerKey {
+        private String tenantId;
+        private String triggerId;
+
+        public TriggerKey(String tenantId, String triggerId) {
+            this.tenantId = tenantId;
+            this.triggerId = triggerId;
+        }
+
+        public String getTenantId() {
+            return tenantId;
+        }
+
+        public void setTenantId(String tenantId) {
+            this.tenantId = tenantId;
+        }
+
+        public String getTriggerId() {
+            return triggerId;
+        }
+
+        public void setTriggerId(String triggerId) {
+            this.triggerId = triggerId;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            TriggerKey that = (TriggerKey) o;
+
+            if (tenantId != null ? !tenantId.equals(that.tenantId) : that.tenantId != null) return false;
+            return triggerId != null ? triggerId.equals(that.triggerId) : that.triggerId == null;
         }
-        if (!refresh) {
-            return;
+
+        @Override
+        public int hashCode() {
+            int result = tenantId != null ? tenantId.hashCode() : 0;
+            result = 31 * result + (triggerId != null ? triggerId.hashCode() : 0);
+            return result;
         }
+    }
+
+    private Set<DistributedEvent> optimizeEvents(Set<DistributedEvent> distEvents) {
+        Map<TriggerKey, Operation> map = new HashMap<>();
+        distEvents.stream().forEach(event -> map.put(new TriggerKey(event.getTenantId(), event.getTriggerId()),
+                event.getOperation()));
+        Set<DistributedEvent> optimizedEvents = new HashSet<>();
+        map.entrySet().stream().forEach(entry -> optimizedEvents.add(new DistributedEvent(entry.getValue(),
+                entry.getKey().getTenantId(), entry.getKey().getTriggerId())));
+        return optimizedEvents;
+    }
```

Comment:
The idea is that in a call of refresh() we might have several events for the same trigger.
For example, first creates the trigger, second update trigger with conditions, another one enables to true, and finally it is removed.
So, effectively the last one is what is needed (on this case for the alerter, others might have a different behaviour, so this logic is placed here for this specific alerter), so for same trigger last operation should prevail.
I am reviewing that there is a pending corner case is that original Set is based on HashSet which doesn't guarantee order, a small change here to use a LinkedHashSet will guarantee this behaviour.